### ISSUE TITLE
Courier font is not rendering correctly on Linux

### ISSUE THREAD

**Issue by deron-dev** [mullvad/mullvad-browser#104]
# Bug Report

## Behavior

[This image/vector graphic](https://raw.githubusercontent.com/todotxt/todo.txt/8d19d67227f0a695303c29159bd7bb5c77a21fdb/description.svg) seems to render incorrectly.

Mullvad:

![image](https://github.com/mullvad/mullvad-browser/assets/24536673/15c2c048-00cb-4ac7-8a8a-eea93691fa51)

Brave:

![image](https://github.com/mullvad/mullvad-browser/assets/24536673/80a78d2d-265c-4edd-b3e6-1c5f5850f287)

Firefox:

![image](https://github.com/mullvad/mullvad-browser/assets/24536673/b5b3abaa-fff2-4323-bbb9-5ffd62ba39d0)

## Version & System Info

![image](https://github.com/mullvad/mullvad-browser/assets/24536673/d786a58a-9d7d-4bc6-bbaa-4ae74fe4590a)

```
OS: Fedora Linux 37 (Xfce) x86_64 
Host: Strix 17 GL703GE 1.0 
Kernel: 6.3.8-100.fc37.x86_64 
DE: Xfce 4.16 
WM: Xfwm4 
```
---

Thanks much to MV & Tor devs for an awesome project :)

**Comment by Thorin-Oakenpants** [comment]
FWIW, renders fine on Windows in TB, TB-alpha, TB v13 test build, MB, FF ...

@terminal-root What do you get in Tor Browser?

**Comment by ruihildt** [comment]
I can reproduce the issue in MB 12.0.7 and TB 12.5 on Fedora 38/KDE.

**Comment by PieroV** [comment]
Hello @terminal-root.
That SVG file uses `Courier`, which isn't a standard font. It should use `Courier,monospace`, instead.
That said, for some other problems we had a while ago, I'd love to make MB for Linux alias Arial, Courier and Times New Roman with Arimo, Cousine and Tinos (see https://gitlab.torproject.org/tpo/applications/tor-browser/-/issues/41799).

**Comment by Thorin-Oakenpants** [comment]
> I'd love to make MB for Linux alias Arial, Courier and Times New Roman

I'd rather not go down this slippery slope

> That SVG file uses Courier, 

So this SVG file also breaks in FF then (assuming the user doesn't have courier). Someone should confirm that. Or inspect a div set as courier and look at the font inspector

We shouldn't be in the business of unbreaking bad website design. The default font (for en-US) is `serif` and this is why it doesn't line up. So this is not TB/MB's fault. The issue title is misleading now - most/all/some linux distro's do not ship with courier

**Comment by ruihildt** [comment]
Technically you're correct.

Though I feel this is a much a technical decision as it is a user experience decision. It seems it's a common method for various Linux OS to substitute those proprietary fonts with matching free alternatives.

The difference in font rendering between OS is an issue, and here the user expects it to work, as it works in Firefox on the same OS.

**Comment by ruihildt** [comment]
I have tested on mine (Fedora/KDE) and I don't have Courier installed, so it's also substituted somehow.

**Comment by PieroV** [comment]
> I have tested on mine (Fedora/KDE) and I don't have Courier installed, so it's also substituted somehow.

Nimbus Mono PS also on Fedora (just tested with Fedora 37 KDE live on a VM).
Not sure 37 is still the latest, in case could you check the fonts tab on the inspector?
It's hidden in the arrow menu.

Also, I've checked.
Courier and Courier New don't have the same metrics, so me talking about Courier New is slightly misleading.
However, I haven't tested MB without fonts.conf.

**Comment by Thorin-Oakenpants** [comment]
FYI: it is broken on FF on android (i.e it is not monospaced). Mac ships with and we whitelist courier.

> Could you please elaborate more?
> Users will have them rendered with Arimo, Cousine and Tinos.

So after percolating in my head as I watched a movie
- you alias the three fonts (why all three/ surely the only relevant one is monospace for cases like this - alignment)
- this needs testing
   - do we then need to whitelist the fonts for the alias mapping to take place? IDK? Maybe not since they were still detcted in out earlier tests (because the sizes changed) - this explains that the missing font.config fell back to the linux system one, the one which mapped them already?
   - if the fonts used are then the same (arimo etc) then there should be no (extra) entropy, but the fonts would be detected as present (which is fine)

This is all terribly exciting, I must say. Such entropy to be gained in the fallback for courier et al, since almost every distro seems to differ on what to use and do. Great Scott!

**Comment by Thorin-Oakenpants** [comment]
Hmm .. so does FF without RFP on linux detect courier when courier is not installed. If you use TZP and look at the sizes, you might fine e.g. Courier and Liberation Mono both detected with the same sizes

**Comment by Thorin-Oakenpants** [comment]
if you want to modify the fntData object via console, go ahead - add add more fonts to `fntData.full` - for linux this already contains courier and liberation mono

**Comment by PieroV** [comment]
>  * if the fonts used are then the same (arimo etc) then there should be no (extra) entropy, but the fonts would be detected as present (which is fine)
> 

Yes, IIRC it never fell back to system fonts.
From what I remember Firefox did this trick: «Uhm, Courier New you say? I sure know it, it's Cousine. Oh wait, Cousine it's allowed! Let's just render it!!».

But as you've seen, my system thinks "Courier New" is "Liberation Mono".
But also:

```
fc-match Cousine
LiberationMono-Regular.ttf: "Liberation Mono" "Regular"
```

I wonder if somehow it did the opposite when Liberation Mono is not present but Cousine is. I need some sandboxing to test this (but maybe I need food first :grin: ).

**Comment by Thorin-Oakenpants** [comment]
Right, that's what I though. Windows does something maybe similar ... `MS Shell Dlg \32` you say .. wait, that's Tahoma (usually AFAICT, MS docs is very vague) - but since this is a system alias (like the linux alises), it just works. Weirdly though, even if not whitelisted, it is still detected (might have something to do with how windows uses widgets?)

In terms of our fallback, it's not really going to work either, because the font config is missing so _our_ fallback is missing - and the system one takes over (if it has one) = so it's still broken

**Comment by ruihildt** [comment]
> Not sure 37 is still the latest, in case could you check the fonts tab on the inspector?

It's DejaVu Sans Mono. I'm on Fedora 38/KDE. (Gnome, the default Fedora interface potentially has different font choice?)

**Comment by PieroV** [comment]
> > Not sure 37 is still the latest, in case could you check the fonts tab on the inspector?
> 
> It's DejaVu Sans Mono. I'm on Fedora 38/KDE. (Gnome, the default Fedora interface potentially has different font choice?)

I was on KDE, too (I have a Fedora KDE iso for when people tell me they have a problem with KDE or with Fedora).
Maybe the difference is between a live system and an installed system.

That makes the situation even worse.

**Comment by Thorin-Oakenpants** [comment]
> That makes the situation even worse

Not in our current setup. If we alias the fonts (and do not whitelist, not sure if this matters that much as long as it works) then we also are not affected (tests pending). And if the font config is not loaded, then there isn't much more we can do - enumeration wise should still be the same, but the measurements may differ. But at least we're still not leaking all the other system fonts (and TZP adds the three fonts to expected). The missing font config exposes other entropy anyway - e.g. ant-aliasing? So the font whitelist is workable as a fallback, but we would want something more robust to detect a broken font config path

**Comment by PieroV** [comment]
We'll very likely resolve this with 14.0.
I opened https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/issues/41237 and sent https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/merge_requests/1036.

**Comment by Thorin-Oakenpants** [comment]
godamnit .. ninja'd by @PieroV .. what is the world coming to

**Comment by ruihildt** [comment]
This has now been fixed with the Mullvad Browser 14.0 release.

See: https://github.com/mullvad/mullvad-browser/releases/tag/14.0
